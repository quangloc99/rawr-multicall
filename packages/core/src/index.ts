export * from './Address';
export * from './InstructionContext';
export * from './instructions';
export * from './Bytes';
export * from './Call';
export * from './buildContract';
export * from './buildRawMulticallContract';
export * from './decodeOutput';
export * from './errors';
export * from './registerPredeployContract';
export * from './Result';
