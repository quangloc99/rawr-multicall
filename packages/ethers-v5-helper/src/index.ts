export * from './types';
export * from './createEthersV5Call';
export * from './error';
