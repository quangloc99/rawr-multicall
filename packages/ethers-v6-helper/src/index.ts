export * from './types';
export * from './error';
export * from './createEthersV6Call';
