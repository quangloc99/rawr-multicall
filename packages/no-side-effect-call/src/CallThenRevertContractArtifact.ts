import * as CallThenRevertContractArtifact from './generated/contracts/CallThenRevert.json';
export { CallThenRevertContractArtifact };

export const CallThenRevertContractByteCode =
    CallThenRevertContractArtifact.contracts['CallThenRevert.yul'].CallThenRevert.evm.bytecode.object;
