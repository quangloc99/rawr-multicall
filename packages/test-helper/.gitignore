artifacts
ethers-v6-contracts/typechain-types