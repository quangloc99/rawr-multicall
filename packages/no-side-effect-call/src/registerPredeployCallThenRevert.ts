import { registerPredeployContract } from '@raw-multicall/core';
import { CallThenRevertContractByteCode } from './CallThenRevertContractArtifact';

export const CallThenRevertContractLabel = '@raw-multicall/no-side-effect-call/CallThenRevertContract';
export function registerPredeployCallThenRevert() {
    registerPredeployContract(CallThenRevertContractLabel, CallThenRevertContractByteCode);
}
