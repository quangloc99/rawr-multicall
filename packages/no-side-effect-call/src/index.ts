import { registerPredeployCallThenRevert } from './registerPredeployCallThenRevert';
registerPredeployCallThenRevert();

export * from './CallThenRevertContractArtifact';
export * from './registerPredeployCallThenRevert';
export * from './wrapNoSideEffectCall';
