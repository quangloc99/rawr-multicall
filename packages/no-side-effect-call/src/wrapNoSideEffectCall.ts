import { Call, labeledAddress, assert, concatBytes, numberToBytes } from '@raw-multicall/core';
import { CallThenRevertContractLabel } from './registerPredeployCallThenRevert';

export function wrapNoSideEffectCall<ReturnType, ErrorType>(
    call: Call<ReturnType, ErrorType>
): Call<ReturnType, ErrorType> {
    return {
        getContractAddress: () => labeledAddress(CallThenRevertContractLabel),
        getData: (context) => {
            const address = context.resolveAddress(call.getContractAddress(context));
            return concatBytes([
                numberToBytes(call.getGasLimit(context) ?? 2 ** 32 - 1, 4),
                address,
                call.getData(context),
            ]);
        },
        decodeOutput: (success, data) => {
            console.log(data);
            assert(!success, 'CallThenRevert contract result should not have successful status');
            const actualSuccess = data[0];
            const actualData = data.slice(1);
            assert(actualSuccess <= 1, 'Actual success must either be 0 or 1');
            return call.decodeOutput(!!actualSuccess, actualData);
        },
        getGasLimit: () => undefined,
        getValue: (context) => call.getValue(context),
        getDependentLabeledContract: (context) => {
            const addr = call.getContractAddress(context);
            const res = call.getDependentLabeledContract?.(context) ?? [];
            if (typeof addr == 'object' && addr.type == 'labeled') return [...res, addr.label];
            return res;
        },
    };
}
