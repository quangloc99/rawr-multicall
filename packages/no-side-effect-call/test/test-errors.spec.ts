import { describeForChain, CHAIN_ID_MAPPING } from '@rawr-multicall/test-helper';
import { ethers } from 'ethers';
import { ThrowError__factory } from '@rawr-multicall/test-helper/ethers-v6-contracts/typechain-types';
import {
    labeledAddress,
    buildRawMulticallContract,
    registerPredeployContract,
    decodeResult,
} from '@rawr-multicall/core';
import { createEthersV6Call } from '@rawr-multicall/ethers-v6-helper';
import { wrapNoSideEffectCall } from '../src';

describeForChain(
    (chain: number) => `Test error for chain ${chain}`,
    (rpcUrl, chain) => {
        const provider = new ethers.JsonRpcProvider(rpcUrl);
        const ThrowErrorFactory = new ThrowError__factory();
        const throwErrorContract = ThrowError__factory.connect(ethers.ZeroAddress);
        const allowPUSH0 = chain !== CHAIN_ID_MAPPING.ARBITRUM;

        it('Test error', async () => {
            const { data: throwErrorContractByteCode } = await ThrowErrorFactory.getDeployTransaction();
            registerPredeployContract('throw-error', throwErrorContractByteCode);
            const calls = [
                wrapNoSideEffectCall(
                    await createEthersV6Call(throwErrorContract, 'justRevert', ['abc'], {
                        withAddress: labeledAddress('throw-error'),
                    })
                ),
                wrapNoSideEffectCall(
                    await createEthersV6Call(throwErrorContract, 'revertError', ['xyz'], {
                        withAddress: labeledAddress('throw-error'),
                    })
                ),
                wrapNoSideEffectCall(
                    await createEthersV6Call(throwErrorContract, 'revertCustom', [1], {
                        withAddress: labeledAddress('throw-error'),
                    })
                ),
                wrapNoSideEffectCall(
                    await createEthersV6Call(throwErrorContract, 'revertPanic', [], {
                        withAddress: labeledAddress('throw-error'),
                    })
                ),
            ] as const;

            const calldata = buildRawMulticallContract(calls, { allowPUSH0 });
            expect(calldata).toMatchSnapshot();
            const res = await provider.call({ data: calldata.byteCode.toString() });
            const decodedRes = decodeResult(calls, res);
            expect(decodedRes).toMatchSnapshot();
        });
    }
);
