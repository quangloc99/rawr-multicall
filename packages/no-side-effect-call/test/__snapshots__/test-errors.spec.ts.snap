// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test error for chain 1 Test error 1`] = `
{
  "byteCode": "0x6106ed6200012a6080397f68c52d0b2b0eb380a3aa3fadad71edba567b36ee672bc5427c362b798bffb9e261055960805ff5507f3d0c80ebc2a7b4730b6d30eb472d6994c2942c1ebeb076118a8107a1101ce78e60446105d95ff55060ff60e0600461076d5f5f607c61061d5f73eefa077c307924b09fd0257ecf88a09f17a80a365af1841b3d841b01815281013d5f823e3d015f5f607c6106995f73eefa077c307924b09fd0257ecf88a09f17a80a365af1841b3d841b01815281013d5f823e3d015f5f603c6107155f73eefa077c307924b09fd0257ecf88a09f17a80a365af1841b3d841b01815281013d5f823e3d015f5f601c6107515f73eefa077c307924b09fd0257ecf88a09f17a80a365af1841b3d841b01815281013d5f823e3d0161076d900361076df3608060405234801561001057600080fd5b50610539806100206000396000f3fe608060405234801561001057600080fd5b506004361061004c5760003560e01c8063082515fd146100515780633e431b121461006f5780637e608c9d1461008b578063d4e6a2b0146100a7575b600080fd5b6100596100c3565b604051610066919061017d565b60405180910390f35b610089600480360381019061008491906102f2565b6100df565b005b6100a560048036038101906100a0919061039b565b6100e8565b005b6100c160048036038101906100bc9190610414565b610127565b005b60008060019050600080826100d89190610470565b9250505090565b80518060208301fd5b81816040517f08c379a000000000000000000000000000000000000000000000000000000000815260040161011e9291906104df565b60405180910390fd5b806040517f110b365500000000000000000000000000000000000000000000000000000000815260040161015b919061017d565b60405180910390fd5b6000819050919050565b61017781610164565b82525050565b6000602082019050610192600083018461016e565b92915050565b6000604051905090565b600080fd5b600080fd5b600080fd5b600080fd5b6000601f19601f8301169050919050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b6101ff826101b6565b810181811067ffffffffffffffff8211171561021e5761021d6101c7565b5b80604052505050565b6000610231610198565b905061023d82826101f6565b919050565b600067ffffffffffffffff82111561025d5761025c6101c7565b5b610266826101b6565b9050602081019050919050565b82818337600083830152505050565b600061029561029084610242565b610227565b9050828152602081018484840111156102b1576102b06101b1565b5b6102bc848285610273565b509392505050565b600082601f8301126102d9576102d86101ac565b5b81356102e9848260208601610282565b91505092915050565b600060208284031215610308576103076101a2565b5b600082013567ffffffffffffffff811115610326576103256101a7565b5b610332848285016102c4565b91505092915050565b600080fd5b600080fd5b60008083601f84011261035b5761035a6101ac565b5b8235905067ffffffffffffffff8111156103785761037761033b565b5b60208301915083600182028301111561039457610393610340565b5b9250929050565b600080602083850312156103b2576103b16101a2565b5b600083013567ffffffffffffffff8111156103d0576103cf6101a7565b5b6103dc85828601610345565b92509250509250929050565b6103f181610164565b81146103fc57600080fd5b50565b60008135905061040e816103e8565b92915050565b60006020828403121561042a576104296101a2565b5b6000610438848285016103ff565b91505092915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601260045260246000fd5b600061047b82610164565b915061048683610164565b92508261049657610495610441565b5b828204905092915050565b600082825260208201905092915050565b60006104be83856104a1565b93506104cb838584610273565b6104d4836101b6565b840190509392505050565b600060208201905081810360008301526104fa8184866104b2565b9050939250505056fea2646970667358221220f95d61a2496eb61076883efe8f4ee8fb4bdcf30517ca2d83028e57e726da18a564736f6c63430008110033603880600c6000396000f3fe6000368180378051819081903660171901906018903490604081901c6001600160a01b03169060e01cf181523d9060203e60013d01601ffdffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa0793e431b12000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000036162630000000000000000000000000000000000000000000000000000000000ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa0797e608c9d0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000378797a0000000000000000000000000000000000000000000000000000000000ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa079d4e6a2b00000000000000000000000000000000000000000000000000000000000000001ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa079082515fd",
  "splittedByteCodes": [
    Uint8Array:0x6106ed,
    Uint8Array:0x6200012a,
    Uint8Array:0x6080,
    Uint8Array:0x39,
    Uint8Array:0x7f68c52d0b2b0eb380a3aa3fadad71edba567b36ee672bc5427c362b798bffb9e2,
    Uint8Array:0x610559,
    Uint8Array:0x6080,
    Uint8Array:0x5f,
    Uint8Array:0xf5,
    Uint8Array:0x50,
    Uint8Array:0x7f3d0c80ebc2a7b4730b6d30eb472d6994c2942c1ebeb076118a8107a1101ce78e,
    Uint8Array:0x6044,
    Uint8Array:0x6105d9,
    Uint8Array:0x5f,
    Uint8Array:0xf5,
    Uint8Array:0x50,
    Uint8Array:0x60ff,
    Uint8Array:0x60e0,
    Uint8Array:0x6004,
    Uint8Array:0x61076d,
    Uint8Array:0x5f,
    Uint8Array:0x5f,
    Uint8Array:0x607c,
    Uint8Array:0x61061d,
    Uint8Array:0x5f,
    Uint8Array:0x73eefa077c307924b09fd0257ecf88a09f17a80a36,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x5f,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x5f,
    Uint8Array:0x5f,
    Uint8Array:0x607c,
    Uint8Array:0x610699,
    Uint8Array:0x5f,
    Uint8Array:0x73eefa077c307924b09fd0257ecf88a09f17a80a36,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x5f,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x5f,
    Uint8Array:0x5f,
    Uint8Array:0x603c,
    Uint8Array:0x610715,
    Uint8Array:0x5f,
    Uint8Array:0x73eefa077c307924b09fd0257ecf88a09f17a80a36,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x5f,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x5f,
    Uint8Array:0x5f,
    Uint8Array:0x601c,
    Uint8Array:0x610751,
    Uint8Array:0x5f,
    Uint8Array:0x73eefa077c307924b09fd0257ecf88a09f17a80a36,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x5f,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x61076d,
    Uint8Array:0x90,
    Uint8Array:0x03,
    Uint8Array:0x61076d,
    Uint8Array:0xf3,
    Uint8Array:0x,
    Uint8Array:0x608060405234801561001057600080fd5b50610539806100206000396000f3fe608060405234801561001057600080fd5b506004361061004c5760003560e01c8063082515fd146100515780633e431b121461006f5780637e608c9d1461008b578063d4e6a2b0146100a7575b600080fd5b6100596100c3565b604051610066919061017d565b60405180910390f35b610089600480360381019061008491906102f2565b6100df565b005b6100a560048036038101906100a0919061039b565b6100e8565b005b6100c160048036038101906100bc9190610414565b610127565b005b60008060019050600080826100d89190610470565b9250505090565b80518060208301fd5b81816040517f08c379a000000000000000000000000000000000000000000000000000000000815260040161011e9291906104df565b60405180910390fd5b806040517f110b365500000000000000000000000000000000000000000000000000000000815260040161015b919061017d565b60405180910390fd5b6000819050919050565b61017781610164565b82525050565b6000602082019050610192600083018461016e565b92915050565b6000604051905090565b600080fd5b600080fd5b600080fd5b600080fd5b6000601f19601f8301169050919050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b6101ff826101b6565b810181811067ffffffffffffffff8211171561021e5761021d6101c7565b5b80604052505050565b6000610231610198565b905061023d82826101f6565b919050565b600067ffffffffffffffff82111561025d5761025c6101c7565b5b610266826101b6565b9050602081019050919050565b82818337600083830152505050565b600061029561029084610242565b610227565b9050828152602081018484840111156102b1576102b06101b1565b5b6102bc848285610273565b509392505050565b600082601f8301126102d9576102d86101ac565b5b81356102e9848260208601610282565b91505092915050565b600060208284031215610308576103076101a2565b5b600082013567ffffffffffffffff811115610326576103256101a7565b5b610332848285016102c4565b91505092915050565b600080fd5b600080fd5b60008083601f84011261035b5761035a6101ac565b5b8235905067ffffffffffffffff8111156103785761037761033b565b5b60208301915083600182028301111561039457610393610340565b5b9250929050565b600080602083850312156103b2576103b16101a2565b5b600083013567ffffffffffffffff8111156103d0576103cf6101a7565b5b6103dc85828601610345565b92509250509250929050565b6103f181610164565b81146103fc57600080fd5b50565b60008135905061040e816103e8565b92915050565b60006020828403121561042a576104296101a2565b5b6000610438848285016103ff565b91505092915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601260045260246000fd5b600061047b82610164565b915061048683610164565b92508261049657610495610441565b5b828204905092915050565b600082825260208201905092915050565b60006104be83856104a1565b93506104cb838584610273565b6104d4836101b6565b840190509392505050565b600060208201905081810360008301526104fa8184866104b2565b9050939250505056fea2646970667358221220f95d61a2496eb61076883efe8f4ee8fb4bdcf30517ca2d83028e57e726da18a564736f6c63430008110033603880600c6000396000f3fe6000368180378051819081903660171901906018903490604081901c6001600160a01b03169060e01cf181523d9060203e60013d01601ffd,
    Uint8Array:0xffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa0793e431b12000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000036162630000000000000000000000000000000000000000000000000000000000ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa0797e608c9d0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000378797a0000000000000000000000000000000000000000000000000000000000ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa079d4e6a2b00000000000000000000000000000000000000000000000000000000000000001ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa079082515fd,
  ],
  "totalValue": 0,
}
`;

exports[`Test error for chain 1 Test error 2`] = `
[
  {
    "error": [Error: No error fragment found for reverted data "0x616263"],
    "success": false,
  },
  {
    "error": [Error: Error("xyz")],
    "success": false,
  },
  {
    "error": [Error: CustomError(1)],
    "success": false,
  },
  {
    "error": [Error: Panic(18)],
    "success": false,
  },
]
`;

exports[`Test error for chain 42161 Test error 1`] = `
{
  "byteCode": "0x6106ed6200013c6080397f68c52d0b2b0eb380a3aa3fadad71edba567b36ee672bc5427c362b798bffb9e261055960806000f5507f3d0c80ebc2a7b4730b6d30eb472d6994c2942c1ebeb076118a8107a1101ce78e60446105d96000f55060ff60e0600461076d60006000607c61061d600073eefa077c307924b09fd0257ecf88a09f17a80a365af1841b3d841b01815281013d6000823e3d0160006000607c610699600073eefa077c307924b09fd0257ecf88a09f17a80a365af1841b3d841b01815281013d6000823e3d0160006000603c610715600073eefa077c307924b09fd0257ecf88a09f17a80a365af1841b3d841b01815281013d6000823e3d0160006000601c610751600073eefa077c307924b09fd0257ecf88a09f17a80a365af1841b3d841b01815281013d6000823e3d0161076d900361076df3608060405234801561001057600080fd5b50610539806100206000396000f3fe608060405234801561001057600080fd5b506004361061004c5760003560e01c8063082515fd146100515780633e431b121461006f5780637e608c9d1461008b578063d4e6a2b0146100a7575b600080fd5b6100596100c3565b604051610066919061017d565b60405180910390f35b610089600480360381019061008491906102f2565b6100df565b005b6100a560048036038101906100a0919061039b565b6100e8565b005b6100c160048036038101906100bc9190610414565b610127565b005b60008060019050600080826100d89190610470565b9250505090565b80518060208301fd5b81816040517f08c379a000000000000000000000000000000000000000000000000000000000815260040161011e9291906104df565b60405180910390fd5b806040517f110b365500000000000000000000000000000000000000000000000000000000815260040161015b919061017d565b60405180910390fd5b6000819050919050565b61017781610164565b82525050565b6000602082019050610192600083018461016e565b92915050565b6000604051905090565b600080fd5b600080fd5b600080fd5b600080fd5b6000601f19601f8301169050919050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b6101ff826101b6565b810181811067ffffffffffffffff8211171561021e5761021d6101c7565b5b80604052505050565b6000610231610198565b905061023d82826101f6565b919050565b600067ffffffffffffffff82111561025d5761025c6101c7565b5b610266826101b6565b9050602081019050919050565b82818337600083830152505050565b600061029561029084610242565b610227565b9050828152602081018484840111156102b1576102b06101b1565b5b6102bc848285610273565b509392505050565b600082601f8301126102d9576102d86101ac565b5b81356102e9848260208601610282565b91505092915050565b600060208284031215610308576103076101a2565b5b600082013567ffffffffffffffff811115610326576103256101a7565b5b610332848285016102c4565b91505092915050565b600080fd5b600080fd5b60008083601f84011261035b5761035a6101ac565b5b8235905067ffffffffffffffff8111156103785761037761033b565b5b60208301915083600182028301111561039457610393610340565b5b9250929050565b600080602083850312156103b2576103b16101a2565b5b600083013567ffffffffffffffff8111156103d0576103cf6101a7565b5b6103dc85828601610345565b92509250509250929050565b6103f181610164565b81146103fc57600080fd5b50565b60008135905061040e816103e8565b92915050565b60006020828403121561042a576104296101a2565b5b6000610438848285016103ff565b91505092915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601260045260246000fd5b600061047b82610164565b915061048683610164565b92508261049657610495610441565b5b828204905092915050565b600082825260208201905092915050565b60006104be83856104a1565b93506104cb838584610273565b6104d4836101b6565b840190509392505050565b600060208201905081810360008301526104fa8184866104b2565b9050939250505056fea2646970667358221220f95d61a2496eb61076883efe8f4ee8fb4bdcf30517ca2d83028e57e726da18a564736f6c63430008110033603880600c6000396000f3fe6000368180378051819081903660171901906018903490604081901c6001600160a01b03169060e01cf181523d9060203e60013d01601ffdffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa0793e431b12000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000036162630000000000000000000000000000000000000000000000000000000000ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa0797e608c9d0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000378797a0000000000000000000000000000000000000000000000000000000000ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa079d4e6a2b00000000000000000000000000000000000000000000000000000000000000001ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa079082515fd",
  "splittedByteCodes": [
    Uint8Array:0x6106ed,
    Uint8Array:0x6200013c,
    Uint8Array:0x6080,
    Uint8Array:0x39,
    Uint8Array:0x7f68c52d0b2b0eb380a3aa3fadad71edba567b36ee672bc5427c362b798bffb9e2,
    Uint8Array:0x610559,
    Uint8Array:0x6080,
    Uint8Array:0x6000,
    Uint8Array:0xf5,
    Uint8Array:0x50,
    Uint8Array:0x7f3d0c80ebc2a7b4730b6d30eb472d6994c2942c1ebeb076118a8107a1101ce78e,
    Uint8Array:0x6044,
    Uint8Array:0x6105d9,
    Uint8Array:0x6000,
    Uint8Array:0xf5,
    Uint8Array:0x50,
    Uint8Array:0x60ff,
    Uint8Array:0x60e0,
    Uint8Array:0x6004,
    Uint8Array:0x61076d,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x607c,
    Uint8Array:0x61061d,
    Uint8Array:0x6000,
    Uint8Array:0x73eefa077c307924b09fd0257ecf88a09f17a80a36,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x607c,
    Uint8Array:0x610699,
    Uint8Array:0x6000,
    Uint8Array:0x73eefa077c307924b09fd0257ecf88a09f17a80a36,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x603c,
    Uint8Array:0x610715,
    Uint8Array:0x6000,
    Uint8Array:0x73eefa077c307924b09fd0257ecf88a09f17a80a36,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x6000,
    Uint8Array:0x6000,
    Uint8Array:0x601c,
    Uint8Array:0x610751,
    Uint8Array:0x6000,
    Uint8Array:0x73eefa077c307924b09fd0257ecf88a09f17a80a36,
    Uint8Array:0x5a,
    Uint8Array:0xf1,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x3d,
    Uint8Array:0x84,
    Uint8Array:0x1b,
    Uint8Array:0x01,
    Uint8Array:0x81,
    Uint8Array:0x52,
    Uint8Array:0x81,
    Uint8Array:0x01,
    Uint8Array:0x3d,
    Uint8Array:0x6000,
    Uint8Array:0x82,
    Uint8Array:0x3e,
    Uint8Array:0x3d,
    Uint8Array:0x01,
    Uint8Array:0x61076d,
    Uint8Array:0x90,
    Uint8Array:0x03,
    Uint8Array:0x61076d,
    Uint8Array:0xf3,
    Uint8Array:0x,
    Uint8Array:0x608060405234801561001057600080fd5b50610539806100206000396000f3fe608060405234801561001057600080fd5b506004361061004c5760003560e01c8063082515fd146100515780633e431b121461006f5780637e608c9d1461008b578063d4e6a2b0146100a7575b600080fd5b6100596100c3565b604051610066919061017d565b60405180910390f35b610089600480360381019061008491906102f2565b6100df565b005b6100a560048036038101906100a0919061039b565b6100e8565b005b6100c160048036038101906100bc9190610414565b610127565b005b60008060019050600080826100d89190610470565b9250505090565b80518060208301fd5b81816040517f08c379a000000000000000000000000000000000000000000000000000000000815260040161011e9291906104df565b60405180910390fd5b806040517f110b365500000000000000000000000000000000000000000000000000000000815260040161015b919061017d565b60405180910390fd5b6000819050919050565b61017781610164565b82525050565b6000602082019050610192600083018461016e565b92915050565b6000604051905090565b600080fd5b600080fd5b600080fd5b600080fd5b6000601f19601f8301169050919050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052604160045260246000fd5b6101ff826101b6565b810181811067ffffffffffffffff8211171561021e5761021d6101c7565b5b80604052505050565b6000610231610198565b905061023d82826101f6565b919050565b600067ffffffffffffffff82111561025d5761025c6101c7565b5b610266826101b6565b9050602081019050919050565b82818337600083830152505050565b600061029561029084610242565b610227565b9050828152602081018484840111156102b1576102b06101b1565b5b6102bc848285610273565b509392505050565b600082601f8301126102d9576102d86101ac565b5b81356102e9848260208601610282565b91505092915050565b600060208284031215610308576103076101a2565b5b600082013567ffffffffffffffff811115610326576103256101a7565b5b610332848285016102c4565b91505092915050565b600080fd5b600080fd5b60008083601f84011261035b5761035a6101ac565b5b8235905067ffffffffffffffff8111156103785761037761033b565b5b60208301915083600182028301111561039457610393610340565b5b9250929050565b600080602083850312156103b2576103b16101a2565b5b600083013567ffffffffffffffff8111156103d0576103cf6101a7565b5b6103dc85828601610345565b92509250509250929050565b6103f181610164565b81146103fc57600080fd5b50565b60008135905061040e816103e8565b92915050565b60006020828403121561042a576104296101a2565b5b6000610438848285016103ff565b91505092915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601260045260246000fd5b600061047b82610164565b915061048683610164565b92508261049657610495610441565b5b828204905092915050565b600082825260208201905092915050565b60006104be83856104a1565b93506104cb838584610273565b6104d4836101b6565b840190509392505050565b600060208201905081810360008301526104fa8184866104b2565b9050939250505056fea2646970667358221220f95d61a2496eb61076883efe8f4ee8fb4bdcf30517ca2d83028e57e726da18a564736f6c63430008110033603880600c6000396000f3fe6000368180378051819081903660171901906018903490604081901c6001600160a01b03169060e01cf181523d9060203e60013d01601ffd,
    Uint8Array:0xffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa0793e431b12000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000036162630000000000000000000000000000000000000000000000000000000000ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa0797e608c9d0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000378797a0000000000000000000000000000000000000000000000000000000000ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa079d4e6a2b00000000000000000000000000000000000000000000000000000000000000001ffffffff6aa51107bc1d691bc9a7f06226c5bd2207eaa079082515fd,
  ],
  "totalValue": 0,
}
`;

exports[`Test error for chain 42161 Test error 2`] = `
[
  {
    "error": [Error: No error fragment found for reverted data "0x616263"],
    "success": false,
  },
  {
    "error": [Error: Error("xyz")],
    "success": false,
  },
  {
    "error": [Error: CustomError(1)],
    "success": false,
  },
  {
    "error": [Error: Panic(18)],
    "success": false,
  },
]
`;
